tests/test_integration_oidc_flow_variants.py

```diff
@@ -110,3 +113,75 @@ def test_return_jwt_from_userinfo_endpoint(
 
     # a user was created
     assert django_user_model.objects.count() == 1
+
+
+@pytest.mark.vcr
+def test_session_refresh(
+    keycloak_config,
+    settings,
+    mock_state_and_nonce,
+    client,
+    django_user_model,
+    vcr,
+    mocker,
+):
+    session = requests.Session()
+
+    keycloak_config.make_users_staff = True
+    keycloak_config.save()
+    settings.MIDDLEWARE = settings.MIDDLEWARE + [
+        "mozilla_django_oidc_db.middleware.SessionRefresh"
+    ]
+    settings.OIDC_RENEW_ID_TOKEN_EXPIRY_SECONDS = 60
+
+    login_url = reverse("login")
+    django_login_response = client.get(login_url)
+    assert django_login_response.status_code == 302
+
+    # simulate login to Keycloak
+    redirect_uri = keycloak_login(django_login_response["Location"], session=session)
+
+    # complete the login flow on our end
+    callback_response = client.get(redirect_uri)
+
+    assert callback_response.status_code == 302
+    assert callback_response["Location"] == "/admin/"
+
+    # a user was created
+    assert django_user_model.objects.count() == 1
+
+    admin_response = client.get("/admin/")
+
+    # User was successfully logged in
+    assert admin_response.status_code == 200
```

Comment:
I think you can clean up a bunch of assertions here that are not relevant for this test. The fact that a user is created after a valid login is already tested and asserted in another test, and removing the asserts makes it much more obvious what is actually being tested in *this* test.

You should always aim for a visual block structure in tests:

```
setup some thing
setup another thing

call the system under test

assert a thing
assert another thing
```